# FIT5196 Task 2 in Assessment 1
#### Student Name: Siyang Feng
#### Student ID: 28246993

Date: 08/03/2018

Version: 2.0

Environment: Python 3.6.2 and Anaconda 4.3.29

Libraries used:
* pandas 0.20.3 (for data frame, included in Anaconda Python 3.6)
* re 2.2.1 (for regular expression, included in Anaconda Python 3.6)
* numpy 1.13.1 (for data format, included in Anaconda Python 3.6)
* pdftables (for extrat data from PDF files)


## Introduction
* you must correctly parse and extract the table;
* existing Python packages (like: pdfminer or pdftables) can be used, however, APIs (like pdftables_api), which requires API keys to push the PDF file to the server in order to get the file parsed, must not be used;
* it is __not required__ to extract the column labels. Except for the first column, which should be named as "Country Name", the other columns should be indexed with integers as shown in the figure;
* if the number followed by a character "x" in the pdf file,  "x" must be dropped in your script;
* your script must be written in a Jupyter notebook named as "pdf.ipynb";
* and the extracted table should be saved in a CSV file named as "health.csv";
* the input file must only be "health.pdf".

## 1. Import libraries
In this subtask, I choose pdftables to solve this problem. before using pdftables. In python3, package *pdfminer3k* should be installed:
```shell
    conda install -c conda-forge pdfminer3k
```
After that, install the pdftables library:
```shell
    pip install pdftables.six
```
The function *get_tables()* in this library will be used to extract data from pdf:
```python
    pdftables.get_tables()
```

In [1]:
import pandas as pd
import numpy as np
import re
from pdftables import get_tables

## 2. Parse PDF File

Open the PDF file and read and get the table using function `open()` and `get_tables()` which comes from pdftables packadge. Then read the table row by row for observation. Table in each page will be created as one table and the output is a aggregation of all these tables. '===' is added under each table for pattern observation in each table. 

In [2]:
# Open the PDF file and read and then, get the table from PDF
pdfFile = 'health.pdf'
pdfObj = open(pdfFile,'rb')
tables = get_tables(pdfObj)

In [3]:
# print the data row by row
for table in tables:
    for row in table:
        print(row)
    print('==================================\n')

['TABLE 3. H', 'EA', 'LT', 'H', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '']
['', '', '', '', '', '', '', '', '', 'I', 'mmuni', 'zation c', 'overage', '(%)', '', '', '', 'Pneumonia', 'Diarrhoea', '', 'Malaria', '']
['', 'Udris', 'se of bnking ervices', 'asic water  (%)', 'Use of bsanitatiservices', 'asic on  (%)', '', '', '', '', '', '', '', '', '', '', 'Protection at birth (PAB) against', 'Care seekingfor children with  symptoms of', 'Treatment with oral rehydra-tion salts', 'Care seeking for children with fever', 'Children sleep-ing under', 'House-holds with at least one']
['Countries', '', '2015', '', '2015', '', 'BCG', 'DTP1β', 'DTP3β', 'polio3', 'MCV1', 'MCV2^', 'HepB3', 'Hib3', 'rota', 'PCV3', 'tetanusλ', '(%)pneumonia', '(ORS) (%)', '(%)', 'ITNs (%)', 'ITN (%)']
['and areas', 'total', 'urban', 'rural', 'totalurban', 'rural', '', '', '', '', '', '2016', '', '', '', '', '', '', '2', '011–2016*', '', '']
['Afghanistan', '63', '89', '53', '3956', '33', '74

By observing the table above, we found out that the first five lines in each table looks like:
```python
['TABLE 3. H', 'EA', 'LT', 'H', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '']
['', '', '', '', '', '', '', '', '', 'I', 'mmuni', 'zation c', 'overage', '(%)', '', '', '', 'Pneumonia', 'Diarrhoea', '', 'Malaria', '']
['', 'Udris', 'se of bnking ervices', 'asic water  (%)', 'Use of bsanitatiservices', 'asic on  (%)', '', '', '', '', '', '', '', '', '', '', 'Protection at birth (PAB) against', 'Care seekingfor children with  symptoms of', 'Treatment with oral rehydra-tion salts', 'Care seeking for children with fever', 'Children sleep-ing under', 'House-holds with at least one']
['Countries', '', '2015', '', '2015', '', 'BCG', 'DTP1β', 'DTP3β', 'polio3', 'MCV1', 'MCV2^', 'HepB3', 'Hib3', 'rota', 'PCV3', 'tetanusλ', '(%)pneumonia', '(ORS) (%)', '(%)', 'ITNs (%)', 'ITN (%)']
['and areas', 'total', 'urban', 'rural', 'totalurban', 'rural', '', '', '', '', '', '2016', '', '', '', '', '', '', '2', '011–2016*', '', '']

```
It seems like top line of PDF in each page and the title line in each table of PDF each page. 
![head](./PDFhead.png)
Thus, the first five lines in each table can be droped. Then, observe the last few lines in each table, there is no interference line.

Thus, the next step is to drop these useless lines in the top of each table.

In the last table, we only need the data above the 'SUMMARY' line. The part is looks like:
```Python
['Zimbabwe679454', '395431959490', '90', '95', '63', '90', '90', '919080', '51', '4150', '9', '48']
['SUMMARY', '', '', '', '', '', '', '', '', '', '', '']
['East Asia and Pacific969491', '778763939794', '93', '93', '87', '90', '38', '89**111', '72**', '47**67**', '–', '–']
```
Thus, we should drop the lines below the 'SUMMARY'. The structure of tables is lists of list. It will be easy to use for loop to drop the lines we don't need and create a new dataset.

In [4]:
for table in tables:
    for row in table[5:]:
        print(row)
    print('=============================\n')

['Afghanistan', '63', '89', '53', '3956', '33', '74', '73', '65', '60', '62', '39', '65', '65', '0', '65', '65', '62', '46', '63', '5', '26']
['Albania', '91', '93', '90', '9898', '97', '99', '99', '98', '98', '96', '98', '98', '98', '0', '98', '92', '70x', '54x', '71x', '–', '–']
['Algeria', '93', '95', '89', '8790', '82', '99', '96', '91', '91', '94', '96', '91', '91', '0', '61', '92', '66', '25', '–', '–', '–']
['Andorra', '100', '100', '100', '100100', '100', '–', '99', '98', '98', '97', '90', '94', '98', '0', '92', '–', '–', '–', '–', '–', '–']
['Angola', '41', '63', '23', '3962', '21', '58', '79', '64', '66', '49', '26', '64', '64', '53', '58', '78', '49', '43', '51', '22', '31']
['Anguilla', '98', '98', '–', '9797', '–', '–', '–', '–', '–', '–', '–', '–', '–', '–', '–', '–', '–', '–', '–', '–', '–']
['Antigua and Barbuda', '97', '–', '–', '88–', '–', '–', '99', '99', '86', '98', '87', '99', '99', '0', '0', '–', '–', '–', '–', '–', '–']
['Argentina', '100', '100', '100', '9595', 

In [5]:
# create a new list to store the usefal data
newData = list()

# select the useful data
for table in tables:
    for row in table[5:]: # drop the first five rows in each table
        if row[0] == '': # drop the row without country
            continue
        if row[0] == 'SUMMARY': # drop the rows under the summary
            break
        newData.append(row) # add each useful list into the new list

print('Useful length :', len(newData))
for t in newData:
    print(t)

Useful length : 202
['Afghanistan', '63', '89', '53', '3956', '33', '74', '73', '65', '60', '62', '39', '65', '65', '0', '65', '65', '62', '46', '63', '5', '26']
['Albania', '91', '93', '90', '9898', '97', '99', '99', '98', '98', '96', '98', '98', '98', '0', '98', '92', '70x', '54x', '71x', '–', '–']
['Algeria', '93', '95', '89', '8790', '82', '99', '96', '91', '91', '94', '96', '91', '91', '0', '61', '92', '66', '25', '–', '–', '–']
['Andorra', '100', '100', '100', '100100', '100', '–', '99', '98', '98', '97', '90', '94', '98', '0', '92', '–', '–', '–', '–', '–', '–']
['Angola', '41', '63', '23', '3962', '21', '58', '79', '64', '66', '49', '26', '64', '64', '53', '58', '78', '49', '43', '51', '22', '31']
['Anguilla', '98', '98', '–', '9797', '–', '–', '–', '–', '–', '–', '–', '–', '–', '–', '–', '–', '–', '–', '–', '–', '–']
['Antigua and Barbuda', '97', '–', '–', '88–', '–', '–', '99', '99', '86', '98', '87', '99', '99', '0', '0', '–', '–', '–', '–', '–', '–']
['Argentina', '100', '1

With observing the new data list, we have merge all the subtable together using `append()` function. 

However, we can find that some items are merged with their neighbors. In the top three pages, there are only items 5 and 6 are merged in each line *('3956' should be '39' and '56')*. However, in the last pages, there are a lot of items merged together. Country name merged with next three items and items 5 to 10 merged together and so on.

In the above three pages, the data record as:
![first three pages](./PDFLinesAbove.PNG)

The parsed table list is:
```python
['Afghanistan', '63', '89', '53', '3956', '33', '74', '73', '65', '60', '62', '39', '65', '65', '0', '65', '65', '62', '46', '63', '5', '26']
['Albania', '91', '93', '90', '9898', '97', '99', '99', '98', '98', '96', '98', '98', '98', '0', '98', '92', '70x', '54x', '71x', '–', '–']
['Algeria', '93', '95', '89', '8790', '82', '99', '96', '91', '91', '94', '96', '91', '91', '0', '61', '92', '66', '25', '–', '–', '–']
['Andorra', '100', '100', '100', '100100', '100', '–', '99', '98', '98', '97', '90', '94', '98', '0', '92', '–', '–', '–', '–', '–', '–']
```
In the last page, the data record as:
![last page data](./PDFlastPage.PNG)

The parsed table list in last page is:
```python
['Viet Nam919291', '789172959696', '95', '99', '95', '96', '96', '0094', '81', '51–', '9', '10']
['Yemen708563', '609044737671', '65', '70', '49', '71', '71', '597170', '34', '2533', '–', '–']
['Zambia618644', '314919999991', '87', '93', '58', '91', '91', '909085', '70', '6475', '41', '68']
['Zimbabwe679454', '395431959490', '90', '95', '63', '90', '90', '919080', '51', '4150', '9', '48']
```

We can found that, all the numbers are in the same range of 0 to 100. This feature could be used in regular expression. Moreover, a item can also be '-'. Thus, I define the regular expression pattern as `r'–|100|[1-9]?\dx?'` to seperate the merged items except country name. Because there are a lot of part which merged together. I have defined a function for seperation.

In [6]:
# define the function to seperate the merged numbers
def divideMerge(ls, index):
    """
    Divide the merged values. All the merged item are numbers between 0 to 100.
    
    Arguments:
    ls -- a list of data which records a line of one country information.
    index -- the index of the input item in ls
    
    Return:
    ls -- the renew ls
    """
    matchLt = re.findall(r'–|100|[1-9]?\dx?', ls[index]) # use regular expression to 
                                                          # find all data match the pattern
    if len(matchLt) > 1:  # there are more than one items merged together
        ls[index : index+1] = matchLt  # renew the ls by replace the checked item with the new list of items
        print(matchLt)
    return ls

In [7]:
# iterate the newData list use for loop
# 1. seperate the merged country
# 2. seperate the merged numbers
# 3. neglect 'x' followed by number
for i in newData:
    # check if the country item are merged with ather values. 
    # If yes, find out the merged items. If not, continue.
    matchStr = re.search(r'^([a-zA-Z\s(\)]+)(100|[0-9]{2}|–)(100|[0-9]{2}|–)(100|[0-9]{2}|–)$', i[0])
    if matchStr != None:  # country is merged.
        i[0:1] = matchStr.group(1), matchStr.group(2), matchStr.group(3), matchStr.group(4)
    
    # check items without country.
    while True:
        if len(i) == 23:  # if there are 23 items in i, no merged item.
            break
        for j in range(1, len(i)): # check each item in i and find out if the item merged.
            i = divideMerge(i, j)  # used the function defined above.
        print(i)
        print('New length :', len(i))
    
    # neglect 'x' followed by number
    matchX = re.findall(r'\d+x', str(i)) # find all the item contains 'x'
    if len(matchX) != 0:
        for j in matchX:
            idx = i.index(j)
            i[idx] = i[idx][:-1] # neglect 'x'
        

['39', '56']
['Afghanistan', '63', '89', '53', '39', '56', '33', '74', '73', '65', '60', '62', '39', '65', '65', '0', '65', '65', '62', '46', '63', '5', '26']
New length : 23
['98', '98']
['Albania', '91', '93', '90', '98', '98', '97', '99', '99', '98', '98', '96', '98', '98', '98', '0', '98', '92', '70x', '54x', '71x', '–', '–']
New length : 23
['87', '90']
['Algeria', '93', '95', '89', '87', '90', '82', '99', '96', '91', '91', '94', '96', '91', '91', '0', '61', '92', '66', '25', '–', '–', '–']
New length : 23
['100', '100']
['Andorra', '100', '100', '100', '100', '100', '100', '–', '99', '98', '98', '97', '90', '94', '98', '0', '92', '–', '–', '–', '–', '–', '–']
New length : 23
['39', '62']
['Angola', '41', '63', '23', '39', '62', '21', '58', '79', '64', '66', '49', '26', '64', '64', '53', '58', '78', '49', '43', '51', '22', '31']
New length : 23
['97', '97']
['Anguilla', '98', '98', '–', '97', '97', '–', '–', '–', '–', '–', '–', '–', '–', '–', '–', '–', '–', '–', '–', '–', '–', '–'

New length : 23
['93', '97']
['Tonga', '100', '100', '100', '93', '97', '92', '80', '83', '78', '80', '84', '85', '78', '78', '0', '0', '–', '–', '–', '64', '–', '–']
New length : 23
['92', '–']
['Trinidad and Tobago', '97', '–', '–', '92', '–', '–', '–', '97', '97', '84', '86', '65', '97', '97', '0', '91', '–', '74x', '–', '–', '–', '–']
New length : 23
['93', '98']
['Tunisia', '94', '100', '83', '93', '98', '83', '95', '99', '98', '98', '96', '97', '98', '98', '0', '0', '96', '60', '65', '–', '–', '–']
New length : 23
['96', '99']
['Turkey', '99', '99', '100', '96', '99', '89', '96', '99', '98', '98', '98', '85', '98', '98', '0', '98', '90', '–', '–', '–', '–', '–']
New length : 23
['97', '94']
['Turkmenistan', '94', '91', '98', '97', '94', '99', '98', '99', '98', '98', '99', '99', '98', '98', '0', '0', '–', '59', '47', '–', '–', '–']
New length : 23
['88', '–']
['Turks and Caicos Islands', '94', '–', '–', '88', '–', '–', '–', '–', '–', '–', '–', '–', '–', '–', '–', '–', '–', '–', '–

In [8]:
# print out the wrangled data
for i in newData:
    print(i)

['Afghanistan', '63', '89', '53', '39', '56', '33', '74', '73', '65', '60', '62', '39', '65', '65', '0', '65', '65', '62', '46', '63', '5', '26']
['Albania', '91', '93', '90', '98', '98', '97', '99', '99', '98', '98', '96', '98', '98', '98', '0', '98', '92', '70', '54', '71', '–', '–']
['Algeria', '93', '95', '89', '87', '90', '82', '99', '96', '91', '91', '94', '96', '91', '91', '0', '61', '92', '66', '25', '–', '–', '–']
['Andorra', '100', '100', '100', '100', '100', '100', '–', '99', '98', '98', '97', '90', '94', '98', '0', '92', '–', '–', '–', '–', '–', '–']
['Angola', '41', '63', '23', '39', '62', '21', '58', '79', '64', '66', '49', '26', '64', '64', '53', '58', '78', '49', '43', '51', '22', '31']
['Anguilla', '98', '98', '–', '97', '97', '–', '–', '–', '–', '–', '–', '–', '–', '–', '–', '–', '–', '–', '–', '–', '–', '–']
['Antigua and Barbuda', '97', '–', '–', '88', '–', '–', '–', '99', '99', '86', '98', '87', '99', '99', '0', '0', '–', '–', '–', '–', '–', '–']
['Argentina', '100

### Check the wrangled result
There are some weak point by using the regular expression pattern `r'–|100|[1-9]?\dx?'`. This pattern can seperate the merged string into several number between 0 to 100. However, it cannot recognize whether the merged value is 'DD' and 'D' or 'D' and 'DD'. If the value is 'DD' and 'D' the pattern will certainly correct. But if the value is 'D' and 'DD', it will misseperate into 'DD' and 'D'. In this part, we will focus on this problem.

__Check problems 1 :__

The merged value has been divided through regular expression. This step is to check if the division gives a correct value. The column[4] and column[5] was merged before. We find that the value in column[4] must in the range between column[5] and column[6].

In [9]:
# Check of the abnormal line
def checkTUR(newData, urbanIndex):
    """
    Check of the abnormal line in total, urban, rural:
    if the total value is smaller than min(urban, rural) or 
    larger than max(urban, rural), the line will be seen as abnormal.
    
    Arguments:
    newData -- the wrangled total data
    urbanIndex -- the index of urban data in each line of newData
    
    Return:
    errorList -- the dictionary {index : item} of obnormal line 
    """
    errorList = dict()
    for index, i in enumerate(newData):
        if '–' not in [i[urbanIndex-1], i[urbanIndex], i[urbanIndex+1]]:
            urban = int(i[urbanIndex])
            rural = int(i[urbanIndex+1])
            total = int(i[urbanIndex-1])
            if total < min(urban, rural) or total > max(urban, rural):
                errorList[str(index)] = i
    return errorList

In [10]:
# check the column[4-5] with urban index 5
col45 = checkTUR(newData, 5)
print('Fault in column[4-5] :', col45)

# check the column[1-3] with ruban index 2
col123 = checkTUR(newData, 2)
print('Fault in column[1-3] :', col123)

Fault in column[4-5] : {'60': ['Ethiopia', '39', '77', '30', '71', '8', '4', '75', '86', '77', '75', '70', '0', '77', '77', '63', '76', '80', '30', '30', '35', '45', '64']}
Fault in column[1-3] : {}


From the check result above, we only find the abnormal value in _Ethiopia_ with the column number 4, 5:
```python
{'60': ['Ethiopia', '39', '77', '30', '71', '8', '4', '75', '86', '77', '75', '70', '0', '77', '77', '63', '76', '80', '30', '30', '35', '45', '64']}
```

And no obnormal value in column number 1-3.

The item 4 and 5 are merged. It may be divided in incorrect format. Merge again and divided.

In [11]:
p = r"^(\d)(\d{2})$"  # difine the pattern into 'd' and 'dd'
for key in col45:
    merge = newData[int(key)][4] + newData[int(key)][5]
    new = re.search(p, merge)
    newData[int(key)][4], newData[int(key)][5] = new.group(1), new.group(2)
    # check the changed value
    if int(newData[int(key)][4]) > min(int(newData[int(key)][5]), int(newData[int(key)][6])) and int(newData[int(key)][4]) < max(int(newData[int(key)][5]), int(newData[int(key)][6])):
        print('No Abnormal :', newData[int(key)])

No Abnormal : ['Ethiopia', '39', '77', '30', '7', '18', '4', '75', '86', '77', '75', '70', '0', '77', '77', '63', '76', '80', '30', '30', '35', '45', '64']


__Check problems 2 :__

Column 17 is merged with column 16 and 15 in last PDF pages. In column 15, There is no unidigit except 0. In column 17, all the value should greater than 50%. From the pattern of regular expression before:
```python
    r'–|100|[1-9]?\dx?'
```
compare the data in PDF, there should be no fault in column 15. Thus, we will focus on column 16 and 17.

In [12]:
faultValue = dict()
for index, i in enumerate(newData):
    if '–'not in i[17] and int(i[17]) < 50:
        print(i)
        faultValue[str(index)] = i

['Venezuela (Bolivarian Republic of)', '97', '99', '86', '95', '98', '72', '99', '98', '84', '82', '88', '53', '84', '84', '47', '77', '5', '72', '38', '–', '–', '–']


Look at the value, the fault is same as the last one. The string (ddd) was divided into dd and d instead of d and dd. The pattern 
```python 
    r"^(\d)(\d{2})$" 
```
can still be used to fix this problem.

In [13]:
for key in faultValue:
    merge = faultValue[key][16] + faultValue[key][17]
    new = re.search(p, merge)
    # check the value.
    if int(new.group(2)) < 50:
        print('index', key, 'still error.')
        continue
    # assign the new value.
    newData[int(key)][16], newData[int(key)][17] = new.group(1), new.group(2)


After check the problem, The list should be changed into dataframe. Define the country name as the index name of each row.
```python
data[countryName] = data[remainingData]
```
In dictionary, this function will define the column name and column values. Thus, we should convert the column and row using function `data.T`.

In [14]:
data = {}
for row in newData:
    data[row[0]] = row[1:]
data = pd.DataFrame(data)
data = data.T
data.head()

0    1    2    3    4    5   6   7   8   9  ...  12  13  14  15  \
Afghanistan   63   89   53   39   56   33  74  73  65  60 ...  65  65   0  65   
Albania       91   93   90   98   98   97  99  99  98  98 ...  98  98   0  98   
Algeria       93   95   89   87   90   82  99  96  91  91 ...  91  91   0  61   
Andorra      100  100  100  100  100  100   –  99  98  98 ...  94  98   0  92   
Angola        41   63   23   39   62   21  58  79  64  66 ...  64  64  53  58   

             16  17  18  19  20  21  
Afghanistan  65  62  46  63   5  26  
Albania      92  70  54  71   –   –  
Algeria      92  66  25   –   –   –  
Andorra       –   –   –   –   –   –  
Angola       78  49  43  51  22  31  

[5 rows x 22 columns]

In [15]:
# replace '-' with empty using re
data.replace(r'^–$', '',regex = True, inplace = True)
data.head(10)

0    1    2    3    4    5   6   7   8   9  ...  12  13  \
Afghanistan           63   89   53   39   56   33  74  73  65  60 ...  65  65   
Albania               91   93   90   98   98   97  99  99  98  98 ...  98  98   
Algeria               93   95   89   87   90   82  99  96  91  91 ...  91  91   
Andorra              100  100  100  100  100  100      99  98  98 ...  94  98   
Angola                41   63   23   39   62   21  58  79  64  66 ...  64  64   
Anguilla              98   98        97   97                      ...           
Antigua and Barbuda   97             88                99  99  86 ...  99  99   
Argentina            100  100  100   95   95   94  92  97  92  87 ...  92  92   
Armenia               99   99   99   92   96   83  99  97  94  96 ...  94  94   
Australia            100  100  100  100                98  94  94 ...  94  94   

                     14  15  16  17  18  19  20  21  
Afghanistan           0  65  65  62  46  63   5  26  
Albania               0  98  92  70  54  71          
Algeria               0  61  92  66  25              
Andorra               0  92                          
Angola               53  58  78  49  43  51  22  31  
Anguilla                                             
Antigua and Barbuda   0   0                          
Argentina            75  82      94  18              
Armenia              94  94      57  37  71          
Australia            87  94                          

[10 rows x 22 columns]

In [16]:
# drop all NAN rows and columns
data = data.dropna(1, how = 'all').dropna(0, how = 'all')
data.head()

0    1    2    3    4    5   6   7   8   9  ...  12  13  14  15  \
Afghanistan   63   89   53   39   56   33  74  73  65  60 ...  65  65   0  65   
Albania       91   93   90   98   98   97  99  99  98  98 ...  98  98   0  98   
Algeria       93   95   89   87   90   82  99  96  91  91 ...  91  91   0  61   
Andorra      100  100  100  100  100  100      99  98  98 ...  94  98   0  92   
Angola        41   63   23   39   62   21  58  79  64  66 ...  64  64  53  58   

             16  17  18  19  20  21  
Afghanistan  65  62  46  63   5  26  
Albania      92  70  54  71          
Algeria      92  66  25              
Andorra                              
Angola       78  49  43  51  22  31  

[5 rows x 22 columns]

In [17]:
# set the index column name as 'Country Name'
data.index.name = 'Country Name'
data.head()

0    1    2    3    4    5   6   7   8   9  ...  12  13  14  \
Country Name                                               ...               
Afghanistan    63   89   53   39   56   33  74  73  65  60 ...  65  65   0   
Albania        91   93   90   98   98   97  99  99  98  98 ...  98  98   0   
Algeria        93   95   89   87   90   82  99  96  91  91 ...  91  91   0   
Andorra       100  100  100  100  100  100      99  98  98 ...  94  98   0   
Angola         41   63   23   39   62   21  58  79  64  66 ...  64  64  53   

              15  16  17  18  19  20  21  
Country Name                              
Afghanistan   65  65  62  46  63   5  26  
Albania       98  92  70  54  71          
Algeria       61  92  66  25              
Andorra       92                          
Angola        58  78  49  43  51  22  31  

[5 rows x 22 columns]

In [18]:
# export dataframe into csv file
data.to_csv('./health.csv', encoding = 'utf-16', sep = '\t')

## 3. Summary
This task measured and understanding of basic PDF file processing techniques in Python programming language.

* __PDF parsing and data extraction.__  `open` PDF file in reading as binary object. Install the packadges `pdfminer3k` and `pdftables.six`. Then, import function `get_tables` from module `pdftables`. Function `get_tables()` is used to parse the pdf tables.

* __Features of pdftables.__ The feature of pdftables is that this tool will read pdf file line by line and store each line as a list. Then each page will be stored as a whole table. Finally, the final output will be a aggregate of several dables. The problem of pdftables is that it easily merge the element in same line together. In this task, the majority work is to seperate the merged items in right way. The key point is find out the suitable regular expression pattern. 

* __Regular expression.__ In this task, regular expression is a key way to seperate the merge items. However, a pattern cannot be definately correct. In some cases, a pattern will used because it can recogenize most of the situation correctly. However, we should also know in what kind of situation, the pattern cannot work. We cannot check data line by line if the PDF is extreme big. Thus, after pattern recognization, it is a good way to figure out the logic of the dataset and use the logic to recognize whether all these are correct. If not, try to fix it.

* __Dataframe manipulation.__ `pandas` packadge is used to convert the wrangled data into dataframe for further cleanning. In this part, `replace` all the '-' with nothing and `dropna` the columns and rows without any records with parameter `how=all`. Set the index name with Country name.

* __Exporting data to specific format.__ The last step is to export data to csv format. There, we use function to_csv(). Encoding format, utf-16, was chose because the default encoding, utf-8, cannot recognize all the character correctly.